### PR TITLE
Introduce splatted indicator components

### LINKED ISSUE #3432: Make `rr.Points3D.indicator` possible
so one can do these kinds of things:
```py
rr.log("example/points", [rr.components.Points3D.indicator,
                          rr.components.Position3DBatch(positions)
```

Same story in Rust.

TODO:
- [x] Python
- [x] Rust
- [x] C++

### PR DESCRIPTION
**Pull request by teh-cmc**
This implements splatted indicator components for all languages, which allows for nicer indicator-component-creation APIs, which in turn allows for some pretty sleek use cases as demo'd in the new `manual_indicator` example.

- Fixes #3432

---

```python
rr.log(
    "points_and_mesh",
    [
        rr.Points3D.indicator(),
        rr.Mesh3D.indicator(),
        rr.components.Position3DBatch([[0.0, 0.0, 0.0], [10.0, 0.0, 0.0], [0.0, 10.0, 0.0]]),
        rr.components.ColorBatch([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]),
        rr.components.RadiusBatch([1.0]),
    ],
)
```

```rust
rec.log_component_batches(
    "points_and_mesh",
    false,
    [
        Points3D::indicator().as_ref() as &dyn ComponentBatch,
        Mesh3D::indicator().as_ref(),
        &[[0.0, 0.0, 0.0], [10.0, 0.0, 0.0], [0.0, 10.0, 0.0]].map(Position3D::from),
        &[[255, 0, 0], [0, 255, 0], [0, 0, 255]].map(|[r, g, b]| Color::from_rgb(r, g, b)),
        &[1.0].map(Radius::from),
    ],
)?;
```

```cpp
std::vector<rr::components::Position3D> positions = {
    {0.0, 0.0, 0.0},
    {10.0, 0.0, 0.0},
    {0.0, 10.0, 0.0},
};
std::vector<rr::components::Color> colors = {
    {255, 0, 0},
    {0, 255, 0},
    {0, 0, 255},
};
std::vector<rr::components::Radius> radii = {1.0};

rec.log_component_batches(
    "points_and_mesh",
    3,
    rr::Points3D::indicator(),
    rr::Mesh3D::indicator(),
    positions,
    colors,
    radii
);
```

image (18)

--- 

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3479) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3479)
- [Docs preview](https://rerun.io/preview/871eb8c6c7b207ad10016888e76805c47a95de27/docs) 
- [Examples preview](https://rerun.io/preview/871eb8c6c7b207ad10016888e76805c47a95de27/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

**Comment by teh-cmc** [comment]
I've went the extra mile in order to make indicators fully static and thus pretty much free in Rust because why not, that's very little extra effort.
Others might or might not want to do something similar in Python & C++ at some point... I'm not going to go into that rabbit hole myself right now :grimacing:

File: crates/re_types_builder/build.rs [hunk 1]
```diff
@@ -29,6 +31,11 @@ fn main() {
         return;
     }
 
+    // Only re-build if source-hash exists
+    if !Path::new(SOURCE_HASH_PATH).exists() {
```
Comment (jleibs) [inline crates/re_types_builder/build.rs:35]: Didn't this change already land in another PR?

File: crates/re_types_builder/src/codegen/cpp/mod.rs [hunk 2]
```diff
@@ -1226,6 +1232,33 @@ fn component_to_data_cell_method(
...
+    hpp_includes: &mut Includes,
+    cpp_includes: &mut Includes,
+) -> Method {
+    hpp_includes.insert_rerun("data_cell.hpp");
+    hpp_includes.insert_rerun("arrow.hpp");
+    hpp_includes.insert_rerun("component_batch.hpp");
+    cpp_includes.insert_rerun("indicator_component.hpp");
+
+    Method {
+        docs: "Creates an `AnonymousComponentBatch` out of the associated indicator component. \
```
Comment (jleibs) [inline crates/re_types_builder/src/codegen/cpp/mod.rs:1246]: This isn't anonymous if it has an indicator, is it?

File: docs/code-examples/manual_indicator.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
+// Log some very simple points.
```
Comment (jleibs) [inline docs/code-examples/manual_indicator.cpp:1]:
```suggestion
// Shows how to manually associate one or more indicator components with arbitrary data
```

File: rerun_py/rerun_sdk/rerun/_baseclasses.py [hunk 4]
```diff
@@ -71,8 +71,21 @@ def __str__(self) -> str:
 
         return s
 
-    def archetype_name(self) -> str:
-        return "rerun.archetypes." + type(self).__name__
+    @classmethod
+    def archetype_name(cls) -> str:
+        return "rerun.archetypes." + cls.__name__
```
Comment (jleibs) [inline rerun_py/rerun_sdk/rerun/_baseclasses.py:76]: Just a note: I ran into an issue earlier where subclassing an archetype causes this to have unexpected behavior. We can file as a separate issue though.

File: rerun_py/rerun_sdk/rerun/_log.py
Comment (jleibs) [inline rerun_py/rerun_sdk/rerun/_log.py:1]: There's a `num_instances` in the docstring for the `IndicatorComponentBatch` that needs to go away.  Also the docstring should move to __init__ as discussed before.

**Review by jleibs — APPROVED** [review]

File: crates/re_types_builder/src/codegen/cpp/mod.rs (hunk 2 above)
Comment (teh-cmc) [reply crates/re_types_builder/src/codegen/cpp/mod.rs:1246]: Not sure what you mean. `Anonymous` is what the cpp SDK calls an erased `ComponentBatch`.

Are you saying you'd like this to return a non-erased `ComponentBatch`, or just that the `Anonymous` prefix sounds weird?

File: rerun_py/rerun_sdk/rerun/_baseclasses.py (hunk 4 above)
Comment (teh-cmc) [reply rerun_py/rerun_sdk/rerun/_baseclasses.py:76]: This wasn't a classmethod until this PR, it was a plain one... that might be why you had a bad surprise earlier?
